File: leetcode/BackTracking/leetcode-0090/90.go
package backtracking

import "sort"

func subsetsWithDup(nums []int) [][]int {
	res := [][]int{}
	if nums == nil || len(nums) == 0 {
		return res
	}

	subset := []int{}
	sort.Ints(nums)
	helper(nums, subset, &res, 0)
	return res
}

func helper(nums, subset []int, res *[][]int, start int) {
	tmp := make([]int, len(subset))
	copy(tmp, subset)
	*res = append(*res, tmp)

	if start >= len(nums) {
		return
	}

	for i := start; i < len(nums); i++ {
		if i > start && nums[i] == nums[i-1] {
			continue
		}
		subset = append(subset, nums[i])
		helper(nums, subset, res, i+1)
		subset = subset[:len(subset)-1]
	}
}

File: leetcode/BackTracking/leetcode-0090/90.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return res;
        }
        Arrays.sort(nums);
        List<Integer> subset = new ArrayList<>();
        helper(nums, res, subset, 0);
        return res;
    }

    private void helper(int[] nums, List<List<Integer>> res, List<Integer> subset, int start) {
        List<Integer> tmp = new ArrayList<>(subset);
        res.add(tmp);

        if (start >= nums.length) {
            return;
        }
        
        for (int i = start; i < nums.length; i++) {
            if (i > start && nums[i] == nums[i - 1]) {
                continue;
            }
            subset.add(nums[i]);
            helper(nums, res, subset, i + 1);
            subset.remove(subset.size() - 1);
        }
    }
}